README.md
# breakdancer [![Version npm](https://img.shields.io/npm/v/breakdancer.svg?style=flat-square)](https://www.npmjs.com/package/breakdancer)[![Build Status](https://img.shields.io/travis/godaddy/breakdancer/master.svg?style=flat-square)](https://travis-ci.org/godaddy/breakdancer)[![Dependencies](https://img.shields.io/david/godaddy/breakdancer.svg?style=flat-square)](https://david-dm.org/godaddy/breakdancer)[![Coverage Status](https://img.shields.io/coveralls/godaddy/breakdancer/master.svg?style=flat-square)](https://coveralls.io/r/godaddy/breakdancer?branch=master)

Breakdancer is a small utility library that helps developing and using ergonomic
breakpoints. Devices and sizes will always change, human ergonomics like palm,
wrist, lap, desk, wall never will.

## Table of Contents

- [Installation](#installation)
- [API](#api)
  - [.width()](#width)
  - [.height()](#height)
  - [.viewport()](#viewport)
  - [.changed()](#changed)
  - [.currently()](#currently)
  - [.compareWidth()](#comparewidth)
  - [.compareHeight()](#compareheight)
  - [.breakpoint](#breakpoint)
- [License](#license)

## Installation

The module is released in the public npm registry and can be installed by
running the following command in your terminal:

```
npm install --save breakdancer
```

Please do note that:

- This module was written from a browser environment so it depends on window and
  document to be available in the global scope. We do fallback to a polyfill if
  those are missing so the module should not break in a Node.js environment.
- The module is written in ES6, so make sure that you environment supports this
  or use a transpiler like [Babel](http://babeljs.io/) to compile the code.

## API

The `Breakdancer` constructor is exported as default interface of this module
and can be imported using:

```js
import Breakdancer from 'breakdancer';
```

The `Breakdancer` constructor accepts 2 arguments which are your specification
ergonomic or just breakpoints and an optional reference to a custom `window`
object which should be used for view port information. The specification can
either be an object where the key of the object is name and the value the
width/height specification or an array of object where the objects have a `name`
property.

```js
var breakpoints = new Breakdancer([
  {
    name: 'wrist',
    width: 320
  },
  {
    name: 'palm',
    width: 800,
    height: 600
  }
]);
```

The `width` and/or `height` that you specified should be seen as a minimum value.
So when you specify 320 as width for wrist that means that 0 - 319 will trigger
as `wrist` breakpoint. If width and height are both set it needs to satisfy
the minim value of height as well before our checks move to the next
breakpoint.

Now that everything is specified we can look at which methods are exposed on the
created instance. In all examples we assume that the code above is used as
example.

### width()

Returns the width of the current view port.

```js
breakpoints.width()  // 390
```

### height()

Returns the height of the current view port.

```js
breakpoints.height(); // 509
```

### viewport()

Returns both the height and width of the current view port as an object.

```js
breakpoints.viewport() // { width: 390, height: 509 }
```

### changed()

Check if the view port and therefor our specified breakpoints have changed since
the last time we've called the `changed` method. When constructing an instance
we automatically check what our current satisfied breakpoint is. This method
returns a boolean as indication if a change has occurred.

```js
breakpoints.changed() // false

.. resize the browser ..

breakpoints.changed() // true
```

The method accepts an optional viewport argument if you want to test a different
viewport configuration. If none is provided it will use the viewport information
from the `.viewport()` method.

### currently()

Calculates what we're currently triggering as view port. It does by iterating
over the specified breakpoints and see if they are satisfying in the current
view port. The name of the triggered breakpoint is returned as string.

```js
breakpoints.currently(); // desktop
```

The method accepts an optional viewport argument if you want to test a different
viewport configuration. If none is provided it will use the viewport information
from the `.viewport()` method.

### .breakpoint

A cached value that is used to see if the breakpoint is changed. This is updated
when the `.changed()` method is called. So it's advised to use `.changed()` and
`.breakpoint` over manually calling the `.currently()` method for figuring out
the current break point.

```js
console.log(breakpoints.breakpoint) // wrist

breakpoints.changed();
console.log(breakpoints.breakpoint) // palm
```

### compare()

Returns the difference between the current window and the given breakpoint in
the given dimension. This can be used to check if the window is "greater" than a
breakpoint. If either the given breakpoint or the given attribute do not exist,
a `TypeError` will be thrown.

```js
var breakpoints = new Breakdancer([{
    name: 'wrist',
    width: 320
  }, {
    name: 'palm',
    width: 800,
    height: 600
}]);

// let's assume the window is 500 px wide and 500 px tall.
console.log(breakpoints.compare('wrist', 'width')) // 180
console.log(breakpoints.compare('palm', 'height')) // -100
```

## License

[MIT](LICENSE)


CHANGELOG.md
# CHANGELOG

- **BREAKING:** Drop `react-native` support
- Update dependencies
- Swap to GitHub Actions for CI

### 4.0.0

- Upgrade to `babel@7`.

### 3.0.2

- Removed react-native from peerDependencies.

### 3.0.1

- [#15] Updated peerDependencies ranges to include React 16.
- Bumped devDependencies and remove unused dependencies.
- Added test coverage.
- Enabled Travis-CI.

### 3.0.0 (2.0.0 was skipped due to previously being published)

- [#7] Moved certain dependencies to devDependencies.

### 1.2.0

- [#14] Implement `compare` method to compare current window size to any breakpoint
- Bump the dependencies.

### 1.1.3

- Include index.native.js in `npm run build`. Adds missing `index.native.js` to the publish output.

### 1.1.2

- [#13] Use separate file for `Breakdancer` base class, allows React Native to resolve `Breakdancer`.

## 1.1.1

- [#11], [#12] Removed `react-native` from peerDependencies.

[#7]: https://github.com/godaddy/breakdancer/pull/7
[#11]: https://github.com/godaddy/breakdancer/pull/11
[#12]: https://github.com/godaddy/breakdancer/pull/12
[#13]: https://github.com/godaddy/breakdancer/pull/13
[#14]: https://github.com/godaddy/breakdancer/pull/14
[#15]: https://github.com/godaddy/breakdancer/pull/14


test.native.js
import NativeDancer from './index.native';
import assume from 'assume';
import { Dimensions } from 'react-native';

it('is exposed as a function', function () {
  assume(NativeDancer).is.a('function');
});

describe('nativeDancer', function () {
  let nativeDancer;
  const specification = [
    {
      name: 'mobile',
      width: 400,
      height: 600
    },

    {
      name: 'tablet',
      width: 1024
    },

    {
      name: 'whatever',
      width: 1409
    }
  ];

  beforeEach(function () {
    nativeDancer = new NativeDancer(specification);
  });

  describe('constructor', function () {
    it('stores the specification as arrays', function () {
      assume(nativeDancer.specification).is.a('array');
      assume(nativeDancer.specification[0].name).equals('mobile');
    });

    it('sets the current breakpoint', function () {
      assume(nativeDancer.breakpoint).is.a('string');
      assume(nativeDancer.breakpoint).is.either(['mobile', 'tablet', 'whatever']);
    });
  });

  describe('#height', function () {
    it('equals to device\'s height', function () {
      var bd = new NativeDancer(specification);
      assume(bd.height()).equals(Dimensions.get('window').height);
    });
  });

  describe('#width', function () {
    it('equals to device\'s width', function () {
      var bd = new NativeDancer(specification);
      assume(bd.width()).equals(Dimensions.get('window').width);
    });
  });

  describe('#changed', function () {
    it('returns a boolean if the breakpoint has changed', function () {
      assume(nativeDancer.changed()).is.false();
      assume(nativeDancer.changed()).is.false();
      assume(nativeDancer.changed()).is.false();

      var viewport = { width: 200, height: 500 };

      assume(nativeDancer.changed(viewport)).is.true();
    });

    it('updates the .breakpoint', function () {
      var viewport = { width: 200, height: 500 };

      assume(nativeDancer.breakpoint).does.not.equal('mobile');
      assume(nativeDancer.changed(viewport)).is.true();
      assume(nativeDancer.breakpoint).equals('mobile');
    });
  });

  describe('#currently', function () {
    it('matches as tablet', function () {
      var viewport = { width: 800, height: 600 };

      assume(nativeDancer.currently(viewport)).equals('tablet');
    });

    it('matches as mobile', function () {
      var viewport = { width: 200, height: 500 };

      assume(nativeDancer.currently(viewport)).equals('mobile');
    });

    it('matches unknown', function () {
      var viewport = { width: 1800, height: 1600 };

      assume(nativeDancer.currently(viewport)).equals('unknown');
    });
  });

  describe('#viewport', function () {
    it('returns the width and height', function () {
      const viewport = nativeDancer.viewport();

      assume(viewport.width).is.above(0);
      assume(viewport.width).equals(nativeDancer.width());

      assume(viewport.height).is.above(0);
      assume(viewport.height).equals(nativeDancer.height());
    });
  });
});
